File: feed/luci-app-invizbox/src/luci/model/cbi/privacy_mode.lua
-- Copyright 2016 InvizBox Ltd
-- https://www.invizbox.com/lic/license.txt

local cbi = require "luci.cbi"
local translate = require "luci.i18n"
local sys = require "luci.sys"

local map = cbi.Map("vpn",translate.translate("Privacy Mode"))
map.anonymous = true
map:chain("wizard")
map:chain("tor")
map:chain("openvpn")

local mode_details = map:section(cbi.NamedSection, "active", "active", "",
        translate.translate("Select your privacy mode:"))
mode_details.addremove = false
mode_details.anonymous = true

local mode = mode_details:option(cbi.ListValue, "mode", translate.translate("Choose Privacy Mode").." :" )
mode.id = "mode"
mode.widget = "radio"
mode.template = "cbi/modelvalue"
mode:value("vpn", "VPN")
mode:value("tor", "Tor")
mode:value("extend", "Wifi Extender")

function map.on_after_commit(self)
    local selected_mode = self:formvalue("cbid.vpn.active.mode")
    local product_name = map.uci:get("wizard", "main", "product_name") or "InvizBox Go"
    if product_name == "InvizBox" then
        if selected_mode == "vpn" then
            map.uci:set("tor", "tor", "enabled", "0")
        else
            map.uci:set("tor", "tor", "enabled", "1")
        end
        map.uci:save("tor")
        map.uci:commit("tor")
        sys.call("/etc/init.d/tor restart")
    end
    if selected_mode == "tor" or selected_mode == "extend" then
        map.uci:set("openvpn", "vpn_0", "enabled", "0")
    else
        map.uci:set("openvpn", "vpn_0", "enabled", "1")
    end
    map.uci:save("openvpn")
    map.uci:commit("openvpn")
    sys.call("/etc/init.d/openvpn restart")
end

return map

File: feed/luci-app-invizbox/htdocs/luci-static/resources/view/invizbox/privacyMode.js
'use strict';
'require view';
'require fs';
'require ui';
'require rpc';
'require uci';
'require form';

const privacyMode = view.extend({
	load: function(){ Promise.all([uci.changes(), uci.load('vpn')]); },
	render: function(){
		let m, s, o;

		m = new form.Map('vpn');
		m.chain('ipsec');
		m.chain('openvpn');
		m.chain('tor');

		s = m.section(form.NamedSection, 'active', 'vpn', 'Privacy Mode');

		o = s.option(form.ListValue, 'mode', 'Privacy Mode', 'Select the Privacy Mode for your InvizBox.');
		o.value('vpn', 'VPN');
		o.value('tor', 'Tor');
		o.value('extend', 'WiFi Extender');
		o.widget = 'radio';

		return m.render();
	},
	handleSaveApply: function(ev, mode) {
		return this.handleSave(ev).then(function() {
			const privacyMode = document.querySelector('input[name="cbid.vpn.active.mode"]:checked').value;
			const currentProtocolId = uci.get('vpn', 'active', 'protocol_id');
			const currentProtocol = currentProtocolId && uci.get('vpn', currentProtocolId, 'vpn_protocol') || '';
			if (privacyMode === 'vpn') {
				if (currentProtocol === 'OpenVPN') {
					uci.set('ipsec', 'general', 'enabled', '0');
					uci.set('ipsec', 'vpn_1', 'enabled', '0');
					uci.set('openvpn', 'vpn_1', 'enabled', '1');
				} else {
					uci.set('ipsec', 'general', 'enabled', '1');
					uci.set('ipsec', 'vpn_1', 'enabled', '1');
					uci.set('openvpn', 'vpn_1', 'enabled', '0');
				}
				uci.set('tor', 'tor', 'enabled', '0');
			} else {
				uci.set('ipsec', 'general', 'enabled', '0');
				uci.set('ipsec', 'vpn_1', 'enabled', '0');
				uci.set('openvpn', 'vpn_1', 'enabled', '0');
				uci.set('tor', 'tor', 'enabled', '1');
			}
      uci.save();
      function displayError(){ classes.ui.changes.displayStatus('warning',
					E('p', 'Something went wrong applying the configuration changes…')); }
			classes.ui.changes.displayStatus('notice spinning', E('p', 'Applying configuration changes…'));
			uci.apply().then(function(){
				Promise.all([
					fs.exec_direct('/etc/init.d/tor', [ 'restart' ]),
					fs.exec_direct('/usr/lib/lua/luci/apply_vpn_config.lua')
				]).then(function(){
					classes.ui.changes.confirm(mode == '0', Date.now() + L.env.apply_rollback * 1000);
				}, displayError);
			}, displayError);
		});
	},
  handleReset: null
});
return privacyMode;
